src/openrct2-ui/windows/TopToolbar.cpp

```diff
@@ -278,3593 +278,3549 @@ static Widget window_top_toolbar_widgets[] = {
+
+                auto sceneryPlaceAction = LargeSceneryPlaceAction(
+                    loc, selectedScenery, gWindowSceneryPrimaryColour, gWindowScenerySecondaryColour,
+                    gWindowSceneryTertiaryColour);
+                sceneryPlaceAction.SetCallback([=](const GameAction* ga, const GameActions::Result* result) {
+                    if (result->Error == GameActions::Status::Ok)
+                    {
+                        OpenRCT2::Audio::Play3D(OpenRCT2::Audio::SoundId::PlaceItem, result->Position);
+                    }
+                    else
+                    {
+                        OpenRCT2::Audio::Play3D(OpenRCT2::Audio::SoundId::Error, { loc.x, loc.y, gSceneryPlaceZ });
+                    }
+                });
+                auto res = GameActions::Execute(&sceneryPlaceAction);
+                break;
             }
-            break;
-        case WIDX_WATER:
-            if (!_landToolBlocked)
-            {
-                WindowTopToolbarWaterToolDrag(screenCoords);
-            }
-            break;
-        case WIDX_SCENERY:
-            if (gWindowSceneryPaintEnabled & 1)
-                WindowTopToolbarSceneryToolDown(screenCoords, w, widgetIndex);
-            if (gWindowSceneryEyedropperEnabled)
-                WindowTopToolbarSceneryToolDown(screenCoords, w, widgetIndex);
-            break;
-#ifdef ENABLE_SCRIPTING
-        default:
-            auto& customTool = OpenRCT2::Scripting::ActiveCustomTool;
-            if (customTool)
+            case SCENERY_TYPE_BANNER:
             {
-                customTool->OnDrag(screenCoords);
+                int32_t z;
+                Direction direction;
+                Sub6E1F34Banner(screenCoords, selectedScenery, gridPos, &z, &direction);
+                if (gridPos.IsNull())
+                    return;
+
+                CoordsXYZD loc{ gridPos, z, direction };
+                auto primaryColour = gWindowSceneryPrimaryColour;
+                auto bannerPlaceAction = BannerPlaceAction(loc, selectedScenery, primaryColour);
+                bannerPlaceAction.SetCallback([=](const GameAction* ga, const GameActions::Result* result) {
+                    if (result->Error == GameActions::Status::Ok)
+                    {
+                        auto data = result->GetData<BannerPlaceActionResult>();
+                        OpenRCT2::Audio::Play3D(OpenRCT2::Audio::SoundId::PlaceItem, result->Position);
+                        ContextOpenDetailWindow(WD_BANNER, data.bannerId.ToUnderlying());
+                    }
+                });
+                GameActions::Execute(&bannerPlaceAction);
+                break;
             }
-            break;
-#endif
+        }
     }
-}
 
-/**
- *
- *  rct2: 0x0066CC5B
- */
-static void WindowTopToolbarToolUp(rct_window* w, WidgetIndex widgetIndex, const ScreenCoordsXY& screenCoords)
-{
-    _landToolBlocked = false;
-    switch (widgetIndex)
+    ClearAction GetClearAction()
```

Comment:
It seems like this function could be a `static` function outside the class, since it doesn't rely on any member variable.